ic-agent/src/hash_tree.rs

```diff
@@ -0,0 +1,396 @@
+    /// The label is not part of this node's tree.
+    Absent,
+
+    /// The label was not found, but could still be somewhere else.
+    Continue,
+
+    /// The label was found. Contains a reference to the [HashTreeNode].
+    Found(&'node HashTreeNode),
+}
+
+/// A Node in the HashTree.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub(crate) enum HashTreeNode {
+    Empty(),
+    Fork(Box<(HashTreeNode, HashTreeNode)>),
+    Labeled(Label, Box<HashTreeNode>),
+    Leaf(Vec<u8>),
+    Pruned(Sha256Digest),
+}
+
+impl HashTreeNode {
+    /// Get the domain separator (byte(|s|) . s).
+    #[inline]
+    pub fn domain_sep(&self, hasher: &mut Sha256) {
+        let domain_sep = match self {
+            HashTreeNode::Empty() => "ic-hashtree-empty",
+            HashTreeNode::Fork(_) => "ic-hashtree-fork",
+            HashTreeNode::Labeled(_, _) => "ic-hashtree-labeled",
+            HashTreeNode::Leaf(_) => "ic-hashtree-leaf",
+            HashTreeNode::Pruned(_) => return,
+        };
+        hasher.update(&domain_sep.len().to_ne_bytes());
+        hasher.update(domain_sep.as_bytes());
+    }
+
+    /// Calculate the digest of this node only.
+    #[inline]
+    pub fn digest(&self) -> Sha256Digest {
+        let mut hasher = Sha256::new();
+        self.domain_sep(&mut hasher);
+
+        match self {
+            HashTreeNode::Empty() => {}
+            HashTreeNode::Fork(nodes) => {
+                hasher.update(&nodes.0.digest());
+                hasher.update(&nodes.1.digest());
+            }
+            HashTreeNode::Labeled(label, node) => {
+                hasher.update(&label.as_bytes());
+                hasher.update(&node.digest());
+            }
+            HashTreeNode::Leaf(bytes) => {
+                hasher.update(bytes);
+            }
+            HashTreeNode::Pruned(digest) => {
+                return *digest;
+            }
+        }
+
+        hasher.finish()
+    }
+
+    /// Lookup a single label, returning a reference to the labeled [HashTreeNode] node if found.
+    ///
+    /// This assumes a sorted hash tree, which is what the spec says the system should
+    /// return. It will stop when it finds a label that's greater than the one being looked
+    /// for.
+    ///
+    /// This function is implemented with flattening in mind, ie. flattening the forks
+    /// is not necessary.
+    fn lookup_label(&self, label: &Label) -> LookupLabelResult {
+        match self {
+            // If this node is a labeled node, check for the name. This assume a
+            HashTreeNode::Labeled(l, node) => match label.cmp(l) {
+                std::cmp::Ordering::Greater => LookupLabelResult::Continue,
+                std::cmp::Ordering::Equal => LookupLabelResult::Found(node.as_ref()),
+                // If this node has a smaller label than the one we're looking for, shortcut
+                // out of this search (sorted tree), we looked too far.
+                std::cmp::Ordering::Less => LookupLabelResult::Absent,
```

Comment:
:angry: Github, just ate this long comment. Second try.

```suggestion
                std::cmp::Ordering::Less => LookupLabelResult::Error,
```
I don’t think you can claim `Absent ` here; the lookup codes doesn't yet do the necessary bookkeeping.

Consider this tree:
```
Fork( Fork(Labeled("a",…), Pruned(…)), Labeled("c", …) )
```
looking up label `b` in this tree should return `Unknown` (after all, the value _might_ be in the `Pruned` tree, can't tell).

I recommend to just return `Unknown` here for now, which is good enough to unblock this project; the algorithm can be refined in a later PR.

If you want to do it right now, here is the way to do it: While you traverse the tree in order, you pass a flag around `prev_item_was_pruned`, initially `false`. If you hit `Pruned`, you turn this to `true`, when you hit `Labeled` to `false`. When hitting a label that’s larger (or the end of the tree) than the searched for, return `Absent` if the flag was `false` and `Unknown` if it was `true`.

Reply:
I fixed this whole function after debugging with your examples. They helped a lot, thanks!

Reply:
For reference, it wasn't Error either. It really _is_ absent, but there is another state which is Unknown when we pass a Pruned on the way. It's just a bunch of edge cases that I fixed this morning essentially.

Reply:
>  While you traverse the tree in order, you pass a flag around prev_item_was_pruned

That's almost what I did. I have another LookupLabelResult value which is Unknown. Absent gets transformed into Unknown if it meets a pruned.

Reply:
That works too, nice work!